### PR TITLE
feat: allow numeric, duration, or percentage as y axis of trend line

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

I was graphing refresh age of dashboards and it was annoying to have to keep thinking about what 19,523 seconds might be.

We already almost had y axis formatting in place as line graphs take a `percentage` prop to convert y axis to a percentage. 

NB that assumes that the number provided is already a percentage. i.e. it doesn't convert value N in the series to its percentage against the max in the series. If 235 is provided it is formatted as `235%`

## Changes

* creates a y axis format type as a literal union of "numeric", "percentage", and "duration"
* adds a select to allow user to choose between those formats

## How did you test this code?

adding developer tests and checking locally

### CODE REVIEW

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (Twixes) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx]: I think this button should be to the left of Chart type, to minimize disruption to user workflows.
Also "y-axis" should be capitalized I think.

File: frontend/src/types.ts
Comment (Twixes) [inline frontend/src/types.ts:1107]: The empty line seems extraneous

File: frontend/src/scenes/insights/yAxisFormat.ts
Comment (Twixes) [inline frontend/src/scenes/insights/yAxisFormat.ts:15]: We do already have a `percentage` util, though that one is intended for values in the 0-1 range. Do you think we'll be able to reuse it by just multiplying with 100?

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (Twixes) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx:165]: Why do we need this? It seems odd that an invalid format would be selectable

**Review by Twixes — COMMENTED** [review]
This is pretty cool and the implementation is sensible. Definitely would like a look from @lottiecoxon or @clarkus before merging though (also, this will need to be incorporated into query editor mocks).

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (pauldambra) [reply frontend/src/scenes/insights/InsightDisplayConfig.tsx]: Hmmm, it's only visible if chart type is line. if it's to the left then if you weren't on type "line" and changed to "line" it would appear to the left of the thing you're focussed on... 

I'm not sure which is better... watching it also makes me think maybe it should be disabled rather than hidden

## disabling the control



## hiding and showing to the left



## hiding and showing to the right

File: frontend/src/types.ts
Comment (pauldambra) [reply frontend/src/types.ts:1107]: It's the size of that type. Makes me want to impose structure. But that's a separate task :)

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (pauldambra) [reply frontend/src/scenes/insights/InsightDisplayConfig.tsx:165]: It's only to coerce the type from maybe string to literal value

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (pauldambra) [reply frontend/src/scenes/insights/InsightDisplayConfig.tsx]: I've left it appearing to the left and being disabled if not applicable

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (kappa90) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx:189]: So, are we resolving to in-line classnames as a general rule? I approve

**Review by kappa90 — APPROVED** [review]
🥇 Approved as long as design approves too

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (pauldambra) [reply frontend/src/scenes/insights/InsightDisplayConfig.tsx:189]: Yeah, I had to fix spacing. So thought I'd fix it properly :)

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (benjackwhite) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx:93]: not sure if `mx-0` is doing anything here

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (benjackwhite) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx:92]: Whilst you're here _if you want_ we could delete the other css code in favor of 
```suggestion
        <div className="flex justify-between items-center flex-wrap">
```

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (benjackwhite) [reply frontend/src/scenes/insights/InsightDisplayConfig.tsx:189]: Not saying we have to but in the case of repeated styles, that could be pulled out to a variable `rowClassNames` to dry it up

**Comment by pauldambra** [comment]
Hmmm. maybe "y axis" isn't the name

E.g. if choosing pie chart we'd still display e.g. "500" but could display "8.33m"

Or on a value bar graph could format the X axis...

In grafana they don't let you specify the axis and just call it "unit

Screenshot 2022-08-02 at 10 18 40

For a value graph we could label it "x-axis format" instead. For a pie chart 🤔 "value format"

* line - linear -> y-axis format
* line - cumulative -> y-axis format
* bar - time -> y-axis format
* bar - value -> x-axis format
* table -> ??? unit
* pie -> unit
* world map -> unit

measurement series?

If there's a decent name we could have this for every type

File: frontend/src/scenes/insights/yAxisFormats.test.ts
Comment (benjackwhite) [inline frontend/src/scenes/insights/yAxisFormats.test.ts:3]: Wooohoo tests. Could we maybe add some for the `formatYAxisValue` as well?

**Review by benjackwhite — CHANGES_REQUESTED** [review]
Small comments. 

One thing that terrifies me as that we have new persisted data in the filters that did not require a backend change....

**Comment by benjackwhite** [comment]
Didn't say in my review but - hell yes to this feature 😍 🥳 🙌

**Comment by pauldambra** [comment]
> One thing that terrifies me as that we have new persisted data in the filters that did not require a backend change....

Yep, that's why I added the test to prove/check that the cache key wasn't changing for every trend by adding this 🤣

File: frontend/src/scenes/insights/yAxisFormats.test.ts
Comment (pauldambra) [reply frontend/src/scenes/insights/yAxisFormats.test.ts:3]: > Not saying we have to but in the case of repeated styles, that could be pulled out to a variable rowClassNames to dry it up

The idea that seemed to be missing a name wasn't so much (IMHO) the list of classes but the idea that a config filter is wrapped in a particular way. So I extracted that... A bit more text but hopefully more helpful to the future traveller

**Comment by pauldambra** [comment]
When creating the graph we call the numeric series an "aggregation" so "aggregation axis format"

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (kappa90) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx:142]: Nice! Reusable components ftw

File: frontend/src/scenes/insights/InsightDisplayConfig.tsx
Comment (kappa90) [inline frontend/src/scenes/insights/InsightDisplayConfig.tsx:163]: I still find this name not self-explanatory, could we ask @joethreepwood for some feedback, or add a tooltip?

**Review by kappa90 — APPROVED** [review]
Looks good to me, I'd just check on the naming (left a comment)

**Review by benjackwhite — APPROVED** [review]
I like it and I think we should get in. Bias for impact. We'll hear from the team before if they think the name is wrong and then we can discuss other names 👍

File: frontend/src/scenes/insights/aggregationAxisFormat.ts
Comment (benjackwhite) [inline frontend/src/scenes/insights/aggregationAxisFormat.ts:12]: sorry one request - can we capitalize this to match all the other selects

**Review by benjackwhite — CHANGES_REQUESTED** [review]

File: frontend/src/scenes/insights/aggregationAxisFormat.ts
Comment (pauldambra) [reply frontend/src/scenes/insights/aggregationAxisFormat.ts:12]: Ugh, what?! No! *deletes PR

**Comment by pauldambra** [comment]
Sense-checking the silliest combo I could think of. 

Showing median session duration as a duration when broken down by session duration

Screenshot 2022-08-02 at 13 02 07

File: frontend/src/types.ts
Comment (Twixes) [reply frontend/src/types.ts]: Oh yeah, I agree about that 😄

**Comment by Twixes** [comment]
Good point around naming. I think it could be confusing if the same (fundamentally) feature were called differently based on viz type, so I vote for just "Unit", or "Value unit".

**Review by benjackwhite — APPROVED** [review]

**Comment by Twixes** [comment]
(I think this is great too, but BTW let's wait for a designer review before merging please)

**Review by clarkus — COMMENTED** [review]
Nice changes! Are there any conditions that limit how the formatting values are enabled? Do all properties support percentage, duration, etc? I was working on an [update](https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=11230%3A49224) to the chart containers that adds a "configuration" button to control formatting or customization. It wouldn't make sense to have a single option within that workflow right now, but eventually it could be used to contain this as well as other options. 

Funnels · Conversion steps

Short term, I'm kind of worried about placement here. We're already seeing cases where things start to stack vertically. Some of these other options could roll into a configuration container to promote the most used options while still making the other options available. "Smoothing" seems like a good candidate to consolidate here. 

Screen Shot 2022-08-02 at 8 29 29 AM

@rcmarron was looking into something like this for session properties in insights. Just mentioning in case he has any overlapping use cases.

**Comment by pauldambra** [comment]
> Are there any conditions that limit how the formatting values are enabled?

@clarkus at the minute the only thing stopping me adding it for all charts is the complexity of the change if I did it :)

I guess maybe we're missing the ability to have more than one aggregation axis too... So if I want to compare seconds since last refresh of a dashboard and page views of that dashboard I can't specify which series has which axis - but that's not a new problem

So long as we keep it "just" a front-end formatting then we can add it to all the chart types I think

-----

Yep, agree that it's getting crowded in that box...

**Comment by pauldambra** [comment]
@clarkus would you be up for me merging as is?

I can follow on with extending to the other chart types and tidying that top area?

**Review by clarkus — APPROVED** [review]
Sure fine to merge. I do think we need a bit more separation between the `chart type` and `y-axis format` inputs. 1–2rems of space should be enough to separate those. We should revisit those chart containers soon because we're getting really cramped with options. It was already like this, so don't think it should block this PR. Thanks for working on this.

**Pull request merged**
